### PR TITLE
Add unit test for flow control

### PR DESCRIPTION
**Pull request by zsy619**
### Describe what this PR does / why we need it
Code path: core/flow/*
Target coverage: 50+% statements.
Fix: #45
### Special notes for reviews
go test -v -race ./... -cover
=== RUN   TestSetAndRemoveTrafficShapingGenerator
--- PASS: TestSetAndRemoveTrafficShapingGenerator (0.00s)
=== RUN   TestIsValidFlowRule
--- PASS: TestIsValidFlowRule (0.00s)
=== RUN   TestFlowRule_ResourceName
=== RUN   TestFlowRule_ResourceName/test1
=== RUN   TestFlowRule_ResourceName/test2
--- PASS: TestFlowRule_ResourceName (0.00s)
    --- PASS: TestFlowRule_ResourceName/test1 (0.00s)
    --- PASS: TestFlowRule_ResourceName/test2 (0.00s)
=== RUN   TestFlowRule_String
=== RUN   TestFlowRule_String/test1
--- PASS: TestFlowRule_String (0.00s)
    --- PASS: TestFlowRule_String/test1 (0.00s)
=== RUN   TestFlowRule_LoadFromJson
--- PASS: TestFlowRule_LoadFromJson (0.00s)
=== RUN   TestDefaultTrafficShapingCalculator_CalculateAllowedTokens
=== RUN   TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test1
=== RUN   TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test2
=== RUN   TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test3
=== RUN   TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test4
=== RUN   TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test5
--- PASS: TestDefaultTrafficShapingCalculator_CalculateAllowedTokens (0.00s)
    --- PASS: TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test1 (0.00s)
    --- PASS: TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test2 (0.00s)
    --- PASS: TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test3 (0.00s)
    --- PASS: TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test4 (0.00s)
    --- PASS: TestDefaultTrafficShapingCalculator_CalculateAllowedTokens/test5 (0.00s)
=== RUN   TestDefaultTrafficShapingChecker_DoCheck
=== RUN   TestDefaultTrafficShapingChecker_DoCheck/test1_0
=== RUN   TestDefaultTrafficShapingChecker_DoCheck/test1_1
=== RUN   TestDefaultTrafficShapingChecker_DoCheck/test2_0
=== RUN   TestDefaultTrafficShapingChecker_DoCheck/test2_1
=== RUN   TestDefaultTrafficShapingChecker_DoCheck/test2_2
--- PASS: TestDefaultTrafficShapingChecker_DoCheck (0.00s)
    --- PASS: TestDefaultTrafficShapingChecker_DoCheck/test1_0 (0.00s)
    --- PASS: TestDefaultTrafficShapingChecker_DoCheck/test1_1 (0.00s)
    --- PASS: TestDefaultTrafficShapingChecker_DoCheck/test2_0 (0.00s)
    --- PASS: TestDefaultTrafficShapingChecker_DoCheck/test2_1 (0.00s)
    --- PASS: TestDefaultTrafficShapingChecker_DoCheck/test2_2 (0.00s)
=== RUN   TestThrottlingChecker_DoCheckNoQueueingSingleThread
--- PASS: TestThrottlingChecker_DoCheckNoQueueingSingleThread (0.21s)
=== RUN   TestThrottlingChecker_DoCheckSingleThread
--- PASS: TestThrottlingChecker_DoCheckSingleThread (0.00s)
=== RUN   TestThrottlingChecker_DoCheckQueueingParallel
--- PASS: TestThrottlingChecker_DoCheckQueueingParallel (0.00s)
=== RUN   TestTrafficShapingController_Rule
=== RUN   TestTrafficShapingController_Rule/test1
--- PASS: TestTrafficShapingController_Rule (0.00s)
    --- PASS: TestTrafficShapingController_Rule/test1 (0.00s)
=== RUN   TestTrafficShapingController_FlowChecker
=== RUN   TestTrafficShapingController_FlowChecker/test1
--- PASS: TestTrafficShapingController_FlowChecker (0.00s)
    --- PASS: TestTrafficShapingController_FlowChecker/test1 (0.00s)
=== RUN   TestTrafficShapingController_PerformChecking
=== RUN   TestTrafficShapingController_PerformChecking/test1
=== RUN   TestTrafficShapingController_PerformChecking/test2
--- PASS: TestTrafficShapingController_PerformChecking (0.00s)
    --- PASS: TestTrafficShapingController_PerformChecking/test1 (0.00s)
    --- PASS: TestTrafficShapingController_PerformChecking/test2 (0.00s)
=== RUN   TestTrafficShapingController_FlowCalculator
=== RUN   TestTrafficShapingController_FlowCalculator/test1
--- PASS: TestTrafficShapingController_FlowCalculator (0.00s)
    --- PASS: TestTrafficShapingController_FlowCalculator/test1 (0.00s)
PASS
coverage: 62.6% of statements
ok      github.com/sentinel-group/sentinel-golang/core/flow     (cached)        coverage: 62.6% of statements

### CODE REVIEW

**Comment by zsy619** [comment]
@louyuting 
出现这个问题，看go.mod也没有github.com/alibaba/sentinel-golang这个引用

go: github.com/alibaba/sentinel-golang/core/flow tested by
	github.com/alibaba/sentinel-golang/core/flow.test imports
	github.com/sentinel-group/sentinel-golang/core/base: github.com/sentinel-group/sentinel-golang@v0.0.0-20200227055908-d7b7e665cbff: parsing go.mod:
	module declares its path as: github.com/alibaba/sentinel-golang
	        but was required as: github.com/sentinel-group/sentinel-golang

File: core/flow/tc_default_test.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package flow
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/sentinel-group/sentinel-golang/core/base"
```
Comment (louyuting) [inline core/flow/tc_default_test.go]: Please import "github.com/alibaba/sentinel-golang/core/base"
the module "github.com/sentinel-group/sentinel-golang/" had been deprecated.

File: core/flow/traffic_shaping_test.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
+package flow
+
+import (
+	"encoding/json"
+	"reflect"
+	"testing"
+
+	"github.com/sentinel-group/sentinel-golang/core/base"
```
Comment (louyuting) [inline core/flow/traffic_shaping_test.go]: Please import "github.com/alibaba/sentinel-golang/core/base"
the module "github.com/sentinel-group/sentinel-golang/" had been deprecated.

**Comment by louyuting** [comment]
Some suggestion:
1. You can run "go test -race -v ./..." on your computer to insure all test pass before push code.

File: core/flow/traffic_shaping_test.go (hunk 2 above)
Comment (zsy619) [reply core/flow/traffic_shaping_test.go]: image

File: core/flow/tc_default_test.go (hunk 1 above)
Comment (zsy619) [reply core/flow/tc_default_test.go]: @louyuting other test file use "github.com/sentinel-group/sentinel-golang" not "github.com/alibaba/sentinel-golang"

??

Comment (zsy619) [reply core/flow/tc_default_test.go]: local modify and test.....
image
go.mod modified....

Comment (louyuting) [reply core/flow/tc_default_test.go]: Yes, we had update the module name as "github.com/alibaba/sentinel-golang".
You should stop using "github.com/sentinel-group/sentinel-golang" in your code

File: core/flow/traffic_shaping_test.go (hunk 2 above)
Comment (louyuting) [reply core/flow/traffic_shaping_test.go]: Please run rebase from github.com/alibaba/sentinel-golang master before test.

Comment (zsy619) [reply core/flow/traffic_shaping_test.go]: ok
package rename

File: core/flow/traffic_shaping_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+				rule:           tt.fields.rule,
+			}
+			if got := traffic.Rule(); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("TrafficShapingController.Rule() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTrafficShapingController_FlowChecker(t *testing.T) {
```
Comment (louyuting) [inline core/flow/traffic_shaping_test.go]: It is not necessary to follow the test table specification completely for some unit tests with few test point, and from my perspective, it is more important to simplify.

File: core/flow/traffic_shaping_test.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		flowChecker    TrafficShapingChecker
+		rule           *FlowRule
+	}
+	type args struct {
+		node         base.StatNode
+		acquireCount uint32
+		flag         int32
+	}
+
+	flowCalculator := NewDefaultTrafficShapingCalculator(99)
```
Comment (louyuting) [inline core/flow/traffic_shaping_test.go]: It's better to mock interface TrafficShapingCalculator and TrafficShapingChecker.
The usage of mock object, you can refer to "github.com/stretchr/testify/mock".

The principle of unit test:
Unit test should ensure that the down layer logic is isolated.

File: core/flow/traffic_shaping_test.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+				rule:           tt.fields.rule,
+			}
+			if got := traffic.PerformChecking(tt.args.node, tt.args.acquireCount, tt.args.flag); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("TrafficShapingController.PerformChecking() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestTrafficShapingController_FlowCalculator(t *testing.T) {
```
Comment (louyuting) [inline core/flow/traffic_shaping_test.go]: It is not necessary to follow the test table specification completely for some unit tests with few test point, and from my perspective, it is more important to simplify.

Comment (louyuting) [reply core/flow/traffic_shaping_test.go]: For Example:
func TestTrafficShapingController_FlowCalculator(t *testing.T) {
        flowCalculator := NewDefaultTrafficShapingCalculator(99)
	traffic := &TrafficShapingController{
		flowCalculator: flowCalculator,
		flowChecker:    nil,
		rule:           nil,
	}
	if got := traffic.FlowCalculator(); !reflect.DeepEqual(got, flowCalculator) {
		t.Errorf("TrafficShapingController.FlowCalculator() = %v, want %v", got, tt.want)
	}
}

Please let me know if you have better idea.
Thanks.

**Comment by sczyh30** [comment]
Hi, any progress?

**Comment by louyuting** [comment]
Closed b/c there was no feedback for a long time.

**Pull request closed without merging**